### PR TITLE
Tokens controller approve reject refactor full

### PR DESCRIPTION
**Pull request by bergarces**
## Description

Stops triggering the acceptance or rejection of the approval from within TokensController. That'll be the responsibility of the caller.

This is part of a series of tickets that aim to refactor and simplify how we await for the confirmation promise.

## Changes

- **BREAKING**: Awaits approval request in `TokensController` instead of relaying on `acceptWatchAsset` and `rejectWatchAsset`, which are being removed.
- **BREAKING**: Changes return value for `watchAsset`.

## References
* Fixes https://github.com/MetaMask/MetaMask-planning/issues/454
* Fixes https://github.com/MetaMask/MetaMask-planning/issues/455

## Checklist

- [X] I've updated the test suite for new or updated code as appropriate
- [X] I've updated documentation for new or updated code as appropriate (note: this will usually be JSDoc)
- [X] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [inline packages/assets-controllers/src/TokensController.ts:670]: I have checked and the contents of `AssetSuggestionResult.suggestedAssetMeta` are not used neither in mobile nor in extension.

As a matter of fact not even the address (`AssetSuggestionResult.result`) is being used, just awaited, so for all we cared we could even have `Promise<void>`.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [inline packages/assets-controllers/src/TokensController.ts:102]: We no longer need to keep the status of the suggested asset. Once it is not pending, it has already been accepted/rejected and we are not sending it with an event that will need to act differently based on it.

We could actually get rid of the whole `suggestedAssets` array if we wanted to right now, it's not adding anything, but I wanted to be safe as we might want in future to restart approvals of pending suggested assets if MetaMask is closed before they are confirmed.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [inline packages/assets-controllers/src/TokensController.ts:617]: If this wasn't already a breaking change I'd have avoided doing this, but the changes are already breaking.

The return value is not used at all by neither extension nor mobile.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [inline packages/assets-controllers/src/TokensController.ts:674]: We don't need to specify that `interactingAddress` is not optional, because `selectedAddress` isn't.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [inline packages/assets-controllers/src/TokensController.ts:628]: The reason behind this callback is getting rid of yet another switch statement after the approval is accepted.

I'm pretty sure ERC721 is coming soon and this way of dealing with it should help.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [inline packages/assets-controllers/src/TokensController.ts:66]: The only place where this object is created is in the method watchAsset and it uses:
```
interactingAddress: interactingAddress || selectedAddress
```
The value of selectedAddress is defined as a string, not string | undefined, so this value is never undefined.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [inline packages/assets-controllers/src/TokensController.ts:123]: No longer need for status if we don't want to persist them after restarts.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [inline packages/assets-controllers/src/TokensController.ts:694]: This error return is maintained by:
- Throwing with what is now the `default` clause if the type is not ERC20 (this should probably be changed to throw using `ethErrors` from `eth-rpc-errors`).
- Letting `validateTokenToWatch` throw freely.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [inline packages/assets-controllers/src/TokensController.ts:710]: All the cases in this switch are maintained (except the return value when it resolves):
- case 1: When it resolves, the return value is now void for the reasons explained in the `watchAsset` signature.
- case 2: When it's rejected, instead of throwing an error with a string we will define a `userRejectedRequest` in the same way it is done for permissions from the FE (https://github.com/MetaMask/metamask-extension/blob/develop/ui/pages/permissions-connect/permissions-connect.component.js#L383)
- case 3: If it fails for any other reason, we are no longer catching the error so it will throw freely.
- default: This cannot happen.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [inline packages/assets-controllers/src/TokensController.ts:1]: This file has undergone chunky refactors. I suggest to review it by looking both at the diff and the final product.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [inline packages/assets-controllers/src/TokensController.ts:785]: Nothing needs to happen now when rejecting since there's no longer status for `suggestedAssets`, so this is dead code.

File: packages/assets-controllers/src/TokensController.test.ts
Comment (bergarces) [inline packages/assets-controllers/src/TokensController.test.ts:1]: The reason so many tests have been removed, is because some of them were testing the `watchAsset` part and others were testing approval/rejection.

That all happens within `watchAsset` now, so they can be simplified maintaining coverage.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [reply packages/assets-controllers/src/TokensController.ts:610]: This makes more sense when looking at the before/after of how it's handled in the clients
- PR for extension https://github.com/MetaMask/metamask-extension/pull/18829/files#diff-f3e25ebd138f61f51874e39dcbfcc24b8aff46248bf5e7deebfc9561a88d3824R39
- Mobile is similar, no result is expected

File: packages/assets-controllers/src/TokensController.ts
Comment (matthewwalsh0) [inline packages/assets-controllers/src/TokensController.ts:165]: Was this never used or have we now stopped using it?

File: packages/assets-controllers/src/TokensController.ts
Comment (matthewwalsh0) [reply packages/assets-controllers/src/TokensController.ts:610]: For reference, the counter argument is that it could be good practice to provide useful return data in all our API methods for the sake of future-proofing. Though given the level of complexity we've reached, I'd agree to simplify where we can as we can add this back when needed.

Naturally this wouldn't be possible if the return value ever reached the dApp, but currently the only result from the middleware is a boolean.

File: packages/assets-controllers/src/TokensController.ts
Comment (matthewwalsh0) [inline packages/assets-controllers/src/TokensController.ts:611]: At a minimum, is this worth putting in an internal constant? Or should we go all out and create an exported enum to use for the `type` argument in order to discourage these errors?

File: packages/assets-controllers/src/TokensController.ts
Comment (matthewwalsh0) [reply packages/assets-controllers/src/TokensController.ts:613]: This sounds good to me, an `if` is cleaner if we don't expect to ever have any more cases.

File: packages/assets-controllers/src/TokensController.ts
Comment (matthewwalsh0) [inline packages/assets-controllers/src/TokensController.ts:759]: Really impressive how little logic is needed when we're not maintaining the state!

File: packages/assets-controllers/src/TokensController.ts
Comment (matthewwalsh0) [reply packages/assets-controllers/src/TokensController.ts:116]: As this was previously persisted, does this warrant a migration to delete this data on upgrade?
Not sure what the best practice is and if we prioritise fewer migrations over dormant data in the user state.

Comment (adonesky1) [reply packages/assets-controllers/src/TokensController.ts:116]: How was this decision come to? Not saying I disagree with it. I'm just curious what the thinking was here.

File: packages/assets-controllers/src/TokensController.test.ts
Comment (matthewwalsh0) [inline packages/assets-controllers/src/TokensController.test.ts:1012]: I appreciate it's a negative test as it's verifying we don't have any error handling logic, but should we have a test to verify that the method throws if the promise from the messenger call rejects?

File: packages/assets-controllers/src/TokensController.ts
Comment (adonesky1) [inline packages/assets-controllers/src/TokensController.ts:627]: Should we wrap this in a try/catch and handle a failure state gracefully since the catch within the method is removed?

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [reply packages/assets-controllers/src/TokensController.ts:610]: The thing is that I'm not entirely sure what would be useful to return in this case. The `suggestedAssetMeta` or the `Token` from `addToken`.

If we do the second, and then we end up adding more types, it's going to turn into a bigger breaking change.

I think leaving it to return void is safest, as we probably won't need to update anything in every client if we eventually return something (even though it still counts as breaking change).

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [reply packages/assets-controllers/src/TokensController.ts:611]: There's a constant already defined for `"ERC20"` in `controller-utils`.

The enum for types sounds good, but we'd first have to look a bit deeper into what is this controller used for, because even though there are mentions to `ERC721`, there's a whole different controller for NFTs.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [reply packages/assets-controllers/src/TokensController.ts:627]: I don't think we should after these changes.

If, for example, the user rejects to add the token from the front end, we now rely on them sending the `userRejectedRequest` error, which we want to throw as it is.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [reply packages/assets-controllers/src/TokensController.ts:165]: We have now stopped using it. This was called from within the catch statements so that we could send the error to an event handler that would reject the promise.

We can now throw directly from `watchAsset`.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [reply packages/assets-controllers/src/TokensController.ts:116]: The ability to persist pending `wallet_watchAsset` after a restart had no tests to cover that case and it was inadvertently broken in a previous PR.

When we found out about the regression, we reached out, asked if that feature was intentional, accidental or important, so that we could decide between fixing it first. The decision was to not restore that feature and have it behave the same way as other confirmations such as the ones that ask specific permissions (which are also not persisted).

Comment (cryptotavares) [reply packages/assets-controllers/src/TokensController.ts:116]: @adonesky1 I have discussed this with @danjm and @bschorchit and we didn't thought of any relevant use case. As so we have decided to standardise the pending `wallet_watchAsset` confirmation behaviour (i.e no persisting confirmations after a restart... just like permissions for example. The only exception are transactions). But is there a use case that you can think for persisting pending approvals to add suggested assets?

Comment (adonesky1) [reply packages/assets-controllers/src/TokensController.ts:116]: I guess the one reason I can think of to keep suggestedAsset state (though we can probably accomplish this by looking at the approval controller state) is that to avoid duplicate suggestions. This is obviously not about persisting state across reloads. We do not currently prevent duplicate suggestions:
Screenshot 2023-05-16 at 1 09 17 PM

But I was thinking maybe we ought to?

File: packages/assets-controllers/src/TokensController.test.ts
Comment (bergarces) [reply packages/assets-controllers/src/TokensController.test.ts:1013]: I can't see what value that'll add as part of a unit test. It's a promise that is being awaited and that is not wrapped in a try/catch, so yes, it will throw.

In an e2e test we can check that, when rejected, it indeed throws a `userRejectedRequest` RPC error.

File: packages/assets-controllers/src/TokensController.ts
Comment (bergarces) [reply packages/assets-controllers/src/TokensController.ts:116]: It'd be easier to rate limit approvals from `wallet_watchAsset` type from `ApprovalController`.

**Comment by adonesky1** [comment]
Just want to highlight the NFT side changes that mirror this refactor: https://github.com/MetaMask/core/pull/1173

File: packages/assets-controllers/src/TokensController.test.ts
Comment (matthewwalsh0) [reply packages/assets-controllers/src/TokensController.test.ts:1013]: The value would be it would assert our assumption that it's not catching the approval errors, what you've said concerning the code is correct but the unit test would enforce that requirement going forward and ensure it's not captured and handled in future without us being aware and explicitly changing our expectations.

**Review by matthewwalsh0 — DISMISSED** [review]

**Review by matthewwalsh0 — APPROVED** [review]

File: packages/assets-controllers/src/TokensController.ts
Comment (tommasini) [reply packages/assets-controllers/src/TokensController.ts]: Hey @bergarces just saw this right now, I think interactingAddress it's different from selectedAddress, Mobile app wise, selectedAddress is the address you have selected on your wallet, but in the app browser wise, the interacted address could be another one connected to a dapp.

**Pull request merged**
